### PR TITLE
[java] ClassWithOnlyPrivateConstructorsShouldBeFinal - fix FP with inner private classes

### PR DESCRIPTION
**Pull request by adangel**
classes.

I found this while trying to run PMD 7 on PMD 7 sources (dogfood).

E.g.:
https://github.com/pmd/pmd/blob/f36b99fbd755fbe2f6f4f8b35a4299a936305e26/pmd-core/src/main/java/net/sourceforge/pmd/cpd/GUI.java#L775

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 36 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/9248a51d06c14d5aba64eb560b4ca1b3/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 45895 violations,<br />
introduces 25631 new violations, 38 new errors and 0 new configuration errors,<br />
removes 182080 violations, 27 errors and 3 configuration errors.<br />
<a href="https://chunk.io/pmd/9248a51d06c14d5aba64eb560b4ca1b3/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java [hunk 1]
```diff
@@ -21,9 +23,10 @@ public ClassWithOnlyPrivateConstructorsShouldBeFinalRule() {
     @Override
     public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
         if (node.isRegularClass()
-            && !node.isFinal()
+            && !node.hasModifiers(JModifier.FINAL)
             && hasOnlyPrivateCtors(node)
-            && hasNoSubclasses(node)) {
+            && hasNoSubclasses(node)
+            && hasNoAbstractMethods(node)) {
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java:29]: Why not `node.isAbstract()`? A class has to be abstract to contain abstract methods

**Comment by jsotuyod** [comment]
I'm unsure excluding abstract classes is the right approach here…

If we had a top level class:

```java
public abstract FooUtils {
    private FooUtils() {
        throw new AssertionError("Do not instantiate utility classes");
    }

…
}
```

I'd still consider this something to be flagged. Here `abstract` is being used as "can't instantiate it", which is incorrect. `abstract` actually means "I expect this to be extended". The right approach would be to have this class be `final` ("can't extend"), with a private constructor ("can't instantiate").

I think this false positive can be addressed in a different way, not by waving abstract classes, but by waving *nested non-static classes*. This is down to the fact that these "private constructors" can actually be invoked within the top level class.

**Comment by adangel** [comment]
Hm... I'm working on it, but I'm regularly loose what this rule should do... so, I'm trying to describe, what should be correct as far as I understand (please correct me...):

The rule [ClassWithOnlyPrivateConstructorsShouldBeFinal](https://pmd.github.io/latest/pmd_rules_java_design.html#classwithonlyprivateconstructorsshouldbefinal) should flag classes, that can't be extended, because it has only private constructors. So far so good. This has nothing to do with a constructor, that prevents instantiation by throwing a exception.

This creates the following test cases:
a) public top level class with only private constructors -> class should be made final
b) public abstract top level class with only private constructors -> this should be flagged. It reminds the developer to decided: either abstract and at least protected constructor or final and not abstract.
c) public final top level class with only private constructors -> ok

Now considering inner classes. There is also #2536 . Let's only consider first inner classes that are visible somewhere else (not private inner classes). The same test cases apply here. For inner classes that are public, package-private or protected - if they only contain private constructors -> class should be made final. But unless they are subclassed in the same compilation unit (then making it final would result in compilation error, so the rule shouldn't suggest that).

Now the very special case private inner classes. These can be used inside the same compilation unit without restriction. So, making a constructor private here does not prevent subclassing. We probably indeed should do nothing there. We could suggest to make the class final - but for other reasons (there are no subclasses) - and therefore it would be a different rule.

I think, I was additionally thrown off now by the sentence in the rule description "... unless the private constructor is invoked by a inner class". These are different things... It should probably be "... unless it is a private inner class". Whether the constructors are invoked or not doesn't matter.

I'll try to update the test cases accordingly....

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java [hunk 2]
```diff
@@ -37,7 +38,7 @@ private boolean hasNoSubclasses(ASTClassOrInterfaceDeclaration klass) {
     }
 
     private boolean doesExtend(ASTAnyTypeDeclaration sub, ASTClassOrInterfaceDeclaration superClass) {
-        return sub != superClass && TypeTestUtil.isA(superClass.getTypeMirror(), sub);
+        return sub != superClass && TypeTestUtil.isA(superClass.getTypeMirror().getErasure(), sub);
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java:41]: For now, I ignore the generic type here. This is most likely wrong, but I don't know where in the type system this problem is. For the test case now (see below) the rule seems to be working.

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java:41]: I think your solution is fine. `Base<Object>` is definitely not a subclass of `Base<E>`, if it were we would have problems resolving types within the body of `Base`. It is however a parameterization of `Base`. Checking the erasure is the most correct way to check this I think.

Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java:41]: I thought, that we would replace the type var "E" by a concrete type at some point.... I was debugging the "capture" methods (e.g. the one called here https://github.com/pmd/pmd/blob/0fb4e08171cbc0c976244d88c26b6c0b5c04d552/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/TypeOps.java#L407), but these seem to only replace wildcards. So `Base<E>` stayed `Base<E>` which is indeed something different than `Base<Object>`.

Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ClassWithOnlyPrivateConstructorsShouldBeFinalRule.java:41]: Well here we only check whether the derived type has some parameterization of the superclass in its supertypes. But we don't need to know *which* parameterization, so using `isSubtype` is enough.

If you want to figure the parameterization of the superclass, you can use `getAsSuper`:
```java
class Base<E> {}
class Derived extends Base<Object> {}

JTypeMirror sub = Derived;
JTypeMirrot sup = Base<E>;

JTypeMirror asSuper = sub.getAsSuper(sup.getSymbol());
// asSuper == Base<Object>
```

This performs type variable replacement along the way. 

Throughout the entire type resolution framework type variables are treated nearly like real (concrete) types. They're not placeholder for other types, and don't just match "any type". This is the job of inference variables.

**Pull request merged**